File: samplesJS/Crypto/src/CryptoSample.cpp
// Required for WinMain
#include <Windows.h>

// Oak Nut include
#include <onut/Crypto.h>
#include <onut/Font.h>
#include <onut/onut.h>
#include <onut/Settings.h>
#include <onut/SpriteBatch.h>
#include <onut/Renderer.h>

void init();
void render();

std::string originalText = "One does nut simply";
std::string based64;
uint32_t hash;
std::string sha1;
std::vector<std::pair<std::string, bool>> emails = {
    {"daivuk@gmail.com", false},
    {"email@example.com", false},
    {"firstname.lastname@example.com", false},
    {"email@subdomain.example.com", false},
    {"firstname+lastname@example.com", false},
    {"email@123.123.123.123", false},
    {"email@[123.123.123.123]", false},
    {"\"email\"@example.com", false},
    {"1234567890@example.com", false},
    {"email@example-one.com", false},
    {"_______@example.com", false},
    {"email@example.name", false},
    {"email@example.museum", false},
    {"email@example.co.jp", false},
    {"firstname-lastname@example.com", false},
    {"plainaddress", false},
    {"#@%%#$@#$@#.com", false},
    {"@example.com", false},
    {"Joe Smith <email@example.com>", false},
    {"email.example.com", false},
    {"email@example@example.com", false},
    {".email@example.com", false},
    {"email.@example.com", false},
    {"email..email@example.com", false},
    {"email@example.com (Joe Smith)", false},
    {"email@example", false},
    {"email@-example.com", false},
    {"email@example.web", false},
    {"email@111.222.333.44444", false},
    {"email@example..com", false},
    {"Abc..123@example.com", false}
};

// Main
int CALLBACK WinMain(HINSTANCE appInstance, HINSTANCE prevInstance, LPSTR cmdLine, int cmdCount)
{
    oSettings->setGameName("Crypto Sample");
    ORun(init, nullptr, render);
}

void init()
{
    // Base 64
    based64 = OEncodeBase64(reinterpret_cast<const uint8_t*>(originalText.c_str()), originalText.size());

    // Simple hash
    hash = OHash(originalText);

    // Sha1
    sha1 = OSha1(originalText);

    // Validate emails
    for (auto& email : emails)
    {
        email.second = OValidateEmail(email.first);
    }
}

void render()
{
    // Clear
    oRenderer->clear(OColorHex(1d232d));

    // Draw Info
    auto pFont = OGetFont("font.fnt");

    oSpriteBatch->begin();

    pFont->draw("Original: ", {10, 10});
    pFont->draw(originalText, {200, 10}, OTopLeft, Color(.7f, 1.f));

    pFont->draw("Simple Hash: ", {10, 30});
    pFont->draw(std::to_string(hash), {200, 30}, OTopLeft, Color(.7f, 1.f));

    pFont->draw("Sha1: ", {10, 50});
    pFont->draw(sha1, {200, 50}, OTopLeft, Color(.7f, 1.f));

    pFont->draw("Shitty email validation: ", {10, 70});
    Vector2 pos(200, 70);
    for (auto& email : emails)
    {
        if (email.second)
        {
            pFont->draw(email.first, pos, OTopLeft, Color(0, 1, 0, 1.f));
        }
        else
        {
            pFont->draw(email.first, pos, OTopLeft, Color(1, 0, 0, 1.f));
        }
        pos.y += 12;
    }

    oSpriteBatch->end();
}

File: samplesJS/Crypto/assets/scripts/CryptoSample.js
var font = getFont("font.fnt");

var originalText = "One does nut simply";
var based64;
var hash;
var sha1;
var emails = [
    { email: "daivuk@gmail.com", isValid: false },
    { email: "email@example.com", isValid: false },
    { email: "firstname.lastname@example.com", isValid: false },
    { email: "email@subdomain.example.com", isValid: false },
    { email: "firstname+lastname@example.com", isValid: false },
    { email: "email@123.123.123.123", isValid: false },
    { email: "email@[123.123.123.123]", isValid: false },
    { email: "\"email\"@example.com", isValid: false },
    { email: "1234567890@example.com", isValid: false },
    { email: "email@example-one.com", isValid: false },
    { email: "_______@example.com", isValid: false },
    { email: "email@example.name", isValid: false },
    { email: "email@example.museum", isValid: false },
    { email: "email@example.co.jp", isValid: false },
    { email: "firstname-lastname@example.com", isValid: false },
    { email: "plainaddress", isValid: false },
    { email: "#@%%#$@#$@#.com", isValid: false },
    { email: "@example.com", isValid: false },
    { email: "Joe Smith <email@example.com>", isValid: false },
    { email: "email.example.com", isValid: false },
    { email: "email@example@example.com", isValid: false },
    { email: ".email@example.com", isValid: false },
    { email: "email.@example.com", isValid: false },
    { email: "email..email@example.com", isValid: false },
    { email: "email@example.com (Joe Smith)", isValid: false },
    { email: "email@example", isValid: false },
    { email: "email@-example.com", isValid: false },
    { email: "email@example.web", isValid: false },
    { email: "email@111.222.333.44444", isValid: false },
    { email: "email@example..com", isValid: false },
    { email: "Abc..123@example.com", isValid: false }
];

// Base 64
based64 = Cryptography.encodeBase64(originalText);

// Simple hash
hash = Cryptography.hash(originalText);

// Sha1
sha1 = Cryptography.sha1(originalText);

// Validate emails
for (var i = 0; i < emails.length; ++i) {
    var e = emails[i];
    e.isValid = Cryptography.validateEmail(e.email);
}

function render() {
    // Clear
    Renderer.clear(Color.fromHexRGB(0x1d232d));

    // Draw Info
    SpriteBatch.begin();

    SpriteBatch.drawText(font, "Original: ", new Vector2(10, 10));
    SpriteBatch.drawText(font, originalText, new Vector2(200, 10), Vector2.TOP_LEFT, new Color(.7, .7, .7, 1));

    SpriteBatch.drawText(font, "Based64: ", new Vector2(10, 30));
    SpriteBatch.drawText(font, based64, new Vector2(200, 30), Vector2.TOP_LEFT, new Color(.7, .7, .7, 1));

    SpriteBatch.drawText(font, "Simple Hash: ", new Vector2(10, 50));
    SpriteBatch.drawText(font, hash, new Vector2(200, 50), Vector2.TOP_LEFT, new Color(.7, .7, .7, 1));

    SpriteBatch.drawText(font, "Sha1: ", new Vector2(10, 70));
    SpriteBatch.drawText(font, sha1, new Vector2(200, 70), Vector2.TOP_LEFT, new Color(.7, .7, .7, 1));

    SpriteBatch.drawText(font, "Shitty email validation: ", new Vector2(10, 90));
    var pos = new Vector2(200, 90);
    for (var i = 0; i < emails.length; ++i) {
        var e = emails[i];
        if (e.isValid) {
            SpriteBatch.drawText(font, e.email, pos, Vector2.TOP_LEFT, new Color(0, 1, 0, 1));
        }
        else {
            SpriteBatch.drawText(font, e.email, pos, Vector2.TOP_LEFT, new Color(1, 0, 0, 1));
        }
        pos.y += 12;
    }

    SpriteBatch.end();
}
